File: 005_game_of_life/components/game_of_life.js
import {loadResource} from "../../static/utils/resources.js";


export default {
    template: "<div></div>",
    props: {
        resource_path: String,
        speed: Number,
        playing: Boolean,
        drawing: String,
        hex_color: String,
    },
    async mounted() {
        await this.$nextTick(); // NOTE: wait for window.path_prefix to be set
        await loadResource(window.path_prefix + `${this.resource_path}/p5/p5.min.js`);

        this.generation_num = 0;
        this.grid = [];
        this.size = 16;
        this.cols = 0;
        this.rows = 0;
        this.padding = [0, 0];

        this.sketch = new p5((sketch) => {
            sketch.setup = function () {
                sketch.createCanvas(sketch.windowWidth, sketch.windowHeight);
            };
            sketch.draw = this.draw;
            sketch.mouseDragged = this.update_drawing;
            sketch.mouseClicked = this.update_drawing;
        }, this.$el);

        // Add event listeners
        window.onresize = this.resize;
        window.onload = this.init_grid_with_default_pattern;
    },
    methods: {
        get_available_size() {
            const headers = document.getElementsByClassName("q-header")
            const header_height = headers.length === 1 ? headers[0].offsetHeight : 0;

            const footers = document.getElementsByClassName("q-footer")
            const footer_height = footers.length === 1 ? footers[0].offsetHeight : 0;

            return {
                width: this.sketch.windowWidth,
                height: this.sketch.windowHeight - header_height - footer_height,
            };
        },
        draw() {
            // FPS is 60, but we increase generation only if
            const must_generate_next_grid = this.playing
                && this.sketch.frameCount % Math.floor(10 / this.speed) === 0;
            if (must_generate_next_grid) {
                this.generate_next_grid();
            }

            const alive_color = this.sketch.color(this.hex_color)

            this.sketch.clear();
            for (let i = 0; i < this.cols; i++) {
                for (let j = 0; j < this.rows; j++) {
                    if (this.grid[i][j] === 0) {
                        this.sketch.fill(255);
                    } else {
                        this.sketch.fill(alive_color);
                    }
                    this.sketch.square(
                        i * this.size + this.padding[0],
                        j * this.size + this.padding[1],
                        this.size
                    );
                }
            }
        },
        update_drawing() {
            const col = Math.floor(this.sketch.mouseX / this.size);
            const row = Math.floor(this.sketch.mouseY / this.size);

            // Check value exists then update
            if ((this.grid[col] || [])[row] !== undefined) {
                this.grid[col][row] = (this.drawing === "pencil") ? 1 : 0;
            }
        },
        count_neighbors(x, y) {
            let sum = 0;
            for (let i = -1; i < 2; i++) {
                for (let j = -1; j < 2; j++) {
                    let nx = (x + i + this.cols) % this.cols;
                    let ny = (y + j + this.rows) % this.rows;
                    sum += this.grid[nx][ny];
                }
            }
            sum -= this.grid[x][y];
            return sum;
        },
        generate_next_grid() {
            let next_grid = [];
            for (let i = 0; i < this.cols; i++) {
                next_grid[i] = [];
                for (let j = 0; j < this.rows; j++) {
                    const neighbors = this.count_neighbors(i, j);
                    const is_alive = this.grid[i][j] === 1;

                    if (is_alive && neighbors < 2) {
                        next_grid[i][j] = 0;
                    } else if (is_alive && (neighbors === 2 || neighbors === 3)) {
                        next_grid[i][j] = 1;
                    } else if (is_alive && neighbors > 3) {
                        next_grid[i][j] = 0;
                    } else if (!is_alive && neighbors === 3) {
                        next_grid[i][j] = 1;
                    } else {
                        next_grid[i][j] = this.grid[i][j];
                    }
                }
            }
            this.grid = next_grid;
            this.generation_num += 1;
            emitEvent("gol__generation_num", this.generation_num);
        },
        init_grid(mode = "0") {
            const available_size = this.get_available_size()
            this.sketch.resizeCanvas(available_size.width, available_size.height);

            this.generation_num = 0;
            emitEvent("gol__generation_num", this.generation_num);

            // Add padding when computing number of cols and rows
            this.cols = Math.floor(available_size.width / this.size) - 2;
            this.rows = Math.floor(available_size.height / this.size) - 2;
            this.padding = [
                (available_size.width - this.cols * this.size) / 2,
                (available_size.height - this.rows * this.size) / 2,
            ]

            // Init array
            const grid = [];
            for (let i = 0; i < this.cols; i++) {
                grid[i] = [];
                for (let j = 0; j < this.rows; j++) {
                    if (mode === "random") {
                        grid[i][j] = Math.floor(this.sketch.random(2))
                    } else if (mode === "resize" && (this.grid[i] || [])[j] !== undefined) {
                        grid[i][j] = this.grid[i][j];
                    } else {
                        grid[i][j] = 0;
                    }
                }
            }
            this.grid = grid;
        },
        init_grid_with_default_pattern() {
            this.init_grid();

            if (this.cols < 2 || this.rows < 2) {
                return;
            }

            // Default pattern
            this.grid[1][0] = 1;
            this.grid[2][1] = 1;
            this.grid[0][2] = 1;
            this.grid[1][2] = 1;
            this.grid[2][2] = 1;
        },
        resize() {
            this.init_grid("resize");
        },
    },
};

File: 005_game_of_life/components/game_of_life.py
from pathlib import Path

from nicegui import ui


class GameOfLife(
    ui.element,
    component="game_of_life.js",
):
    def __init__(self):
        # A 'Game Of Life' canvas
        super().__init__()
        self.add_resource(Path(__file__).parent.parent / 'libs')
        self._props['speed']: float = 1.0
        self._props['playing']: bool = False
        self._props['drawing']: str = "pencil"
        self._props['hex_color']: str = "#5893a8"

        # Add event listener
        self.generation_num = 0
        ui.on('gol__generation_num', lambda e: setattr(self, "generation_num", e.args))

    def reset(self, random: bool = False, *args, **kwarg):
        self.drawing = "pencil"
        self.run_method("reset", random)

    @property
    def speed(self):
        return self._props['speed']

    @speed.setter
    def speed(self, value):
        self._props['speed'] = min(4., max(0.25, float(value)))
        self.update()

    def decrease_speed(self, *args, **kwarg):
        self.speed /= 2

    def increase_speed(self, *args, **kwarg):
        self.speed *= 2

    @property
    def playing(self):
        return self._props['playing']

    @playing.setter
    def playing(self, value: bool):
        self._props['playing'] = value
        self.update()

    def toggle_play(self, *args, **kwarg):
        self.playing = not self.playing

    def generate_next_grid(self, *args, **kwarg):
        self.playing = False
        self.run_method("generate_next_grid")

    @property
    def drawing(self):
        return self._props['drawing']

    @drawing.setter
    def drawing(self, value: bool):
        self._props['drawing'] = value
        self.update()

    def use_pencil(self, *args, **kwarg):
        self.drawing = "pencil"

    def use_eraser(self, *args, **kwarg):
        self.drawing = "eraser"

    @property
    def hex_color(self):
        return self._props['hex_color']

    @hex_color.setter
    def hex_color(self, value):
        self._props['hex_color'] = value
        self.update()

    def set_hex_color(self, hex_color: str = None):
        self.hex_color = hex_color or "#5893a8"  # the default color
